### DISCUSSION TITLE
deepMove item of array into another path from same obj/array

### DISCUSSION THREAD

**Discussion by crapthings** [YuriGor/deepdash#107] [Ideas]
any possible to provide such feature?

**Comment by crapthings** [comment]
better combine 'obj path' with _.index

deepMove(obj, '[0].a.[3]', '[1].b.[6]')

**Comment by crapthings** [comment]
similar to deepRemove / deepCopy  but all can action by path?

deepRemove(obj, '[0].a.[3]')

deepCopy(obj, '[0].a.[3]', '[1].b.[6]')

**Comment by YuriGor** [comment]
Let me please clarify, do you want this?

_.set(obj, targetPath, _.get(obj, sourcePath));

**Comment by YuriGor** [comment]
Or this?

_.condenseDeep(_.set(obj, targetPath, _.get(obj, sourcePath)));

Note, you should decide what will you do if target path already exists.

**Comment by crapthings** [comment]
https://github.com/rhalff/dot-object#move-a-property-within-one-object-to-another-location

something like this, but it lacks array support i think.

**Comment by YuriGor** [comment]
Could you please describe your use cases?

Also how do you see this method should behave in  case of:
1. target is object and property exists already by the given path
```js
var res = _.moveDeep({a:1,b:2}, 'a', 'b');
```
2. target is array and item exists already  by the given path
```js
var res = _.moveDeep({a:1,b:[2]}, 'a', 'b[0]');
```
3. target is an array and target path points to the not allocated area of array
```js
var res = _.moveDeep({a:1,b:[2]}, 'a', 'b[3]');
```
4. target path go through the not allocated area of some parent array
```js
var res = _.moveDeep({a:1,b:[2]}, 'a', 'b[3][2].c');
```
5. target path points to the child
```js
var res = _.moveDeep({a: {b: {c: 1}}}, 'a', 'a.b.c');
```

**Comment by crapthings** [comment]
maybe move a card from kanban?

the data structure is like

```js
[
  {
    randomid,
    type: 'deck',
    children: [
      {
        randomid,
        type: 'group',
        children: [
          {
            randomid,
            type: 'card',
            name: 'card 1',
          },

          {
            randomid,
            type: 'card',
            name: 'card 2',
          },

          {
            randomid,
            type: 'card',
            name: 'card 3',
          },
        ]
      },

      {
        randomid,
        type: 'card',
        name: 'card 7'
      }
    ]
  },

  {
    randomid,
    type: 'deck',
    children: [
      {
        randomid,
        type: 'card',
        name: 'card 4',
      },

      {
        randomid,
        type: 'card',
        name: 'card 5',
      },

      {
        randomid,
        type: 'card',
        name: 'card 6',
      },
    ]
  },
]
```

like move 'card 5' into 0.children.1.children.2

**Comment by crapthings** [comment]
these deep method is powerful

https://github.com/YuriGor/deepdash#mapdeep
https://github.com/YuriGor/deepdash#index

when you use these with react, mobx

you render the tree deeply and mutate mobx state every thing change,

but sometimes we need something like swap, move, remove kv pair from nested obj or remove element from nested array etc...

**Comment by YuriGor** [comment]
I think you requested too complicated method with unclear specs.
Please think how would you split it Into separate simpler methods.
As for me I'd prefer to do such things manually using lodash get, set and native array methods. If you have some repeating usage patterns - it's better to create helper module, or even a separate package related to specific mobx tasks.
Thank you for interest, please feel free to open new requests.

**Comment by YuriGor** [comment]
Just faced with a similar use case and implemented like this:
- take value by `fromPath`
- delete value at `fromPath`
- if `toPath` points to object - put the value there. Developer cares about possible overwriting by himself.
- if `toPath` points to the array - **insert** value there, **before** possible existing array element, so the value will take a position by given index and possible consequent elements will be shifted right. If given index is greater then the length of the array - empty array slots will be preserved.
- if `fromPath` points to the array - optionally (by default) condense source array to fill the hole possibly left after deletion.
- if `fromPath` and `toPath` are both pointing to the same array - then it means that by default target array will be condensed.
- if `fromPath` does not exist - `undefined` value will be used without error similar to Lodash `get`
- if `toPath` does not exist - it will be created similarly as it will be created by Lodash `set` method.
- if `toPath` points to the depths of `fromPath` - it is the same case when `toPath` does not exist, because `fromPath` will be deleted.

So I am going to implement `moveDeep` as described above.
@crapthings, @FunkySamuel37 could you please review the suggested algorithm, if it fits your practice and covers possible edge cases?
